packages/api/src/controllers/access-control.ts

```diff
@@ -50,73 +57,184 @@ app.post(
         );
-      }
-
-      const signingKey = signingKeyOutput[0];
+    }
+  }
+);
 
-      if (signingKey.userId !== content.userId) {
-        console.log(`
-          access-control: gate: disallowing playback for contentId=${content.id} with playbackId=${playbackId} the content and the public key pub=${req.body.pub} do not share the same owner, signingKeyUserId=${signingKey.userId}, contentUserId=${content.userId}
-        `);
-        throw new NotFoundError("Content not found");
-      }
+async function fireGateWebhook(
+  webhook: DBWebhook,
+  plabackPolicy: PlaybackPolicy,
+  accessKey: string
+) {
+  let timestamp = Date.now();
+  let params = {
+    method: "POST",
+    headers: {
+      "content-type": "application/json",
+      "user-agent": "livepeer.studio",
+    },
+    timeout: WEBHOOK_TIMEOUT,
+    body: JSON.stringify({
+      context: plabackPolicy.webhookContext,
+      accessKey: accessKey,
+    }),
+  };
+
+  if (webhook.sharedSecret) {
+    params = await addSignatureHeader(params, webhook, timestamp);
+  }
 
-      if (signingKey.disabled || signingKey.deleted) {
-        const signingKeyLog = JSON.stringify(JSON.stringify(signingKey));
-        console.log(`
-          access-control: gate: disallowing playback for contentId=${content.id} with playbackId=${playbackId} the public key pub=${signingKey.id} is disabled or deleted, signingKey=${signingKeyLog}
-        `);
-        throw new ForbiddenError("The public key is disabled or deleted");
-      }
-
-      tracking.recordSigningKeyValidation(signingKey.id);
-      res.set("Cache-Control", "max-age=120,stale-while-revalidate=600");
-      res.status(204);
-      return res.end();
-    } else {
-      console.log(`
-        access-control: gate: content with playbackId ${playbackId} is gated but playbackPolicyType ${playbackPolicyType} is not supported, disallowing playback
-      `);
-      throw new BadRequestError(
-        `unknown playbackPolicy type: ${playbackPolicyType}`
-      );
+  const startTime = process.hrtime();
+  let resp: Response;
+  let errorMessage: string;
+  let responseBody: string;
+  let statusCode: number;
+
+  try {
+    console.log(`webhook ${webhook.id} firing`);
+    resp = await fetchWithTimeout(webhook.url, params);
+    responseBody = await resp.text();
+    statusCode = resp.status;
+
+    if (resp.status >= 200 && resp.status < 300) {
+      // 2xx requests are cool. all is good
+      console.log(`webhook ${webhook.id} fired successfully`);
     }
+
+    console.error(
+      `webhook ${webhook.id} didn't get 200 back! response status: ${resp.status}`
+    );
+  } catch (e) {
+    console.log(`
+      webhook ${webhook.id} failed to fire: ${e.message}
+    `);
+    errorMessage = e.message;
+    statusCode = 0;
+  } finally {
+    await this.storeTriggerStatus(
+      webhook,
+      timestamp,
+      statusCode,
+      errorMessage,
+      responseBody
+    );
+    return statusCode;
```

Comment:
This return should be either inside the `try` or after the `finally`, you should almost never return from a `finally` block. Not only the data flow can get confusing, it can also lead to really unexpected results.
e.g. Any exception thrown in the try block will be eaten by this. If you change line 227 to throw an exception instead, it wouldn't work and likely an `undefined` status code would be returned instead lol